versioning/worker-versions.md

```diff
@@ -0,0 +1,174 @@
+  attempted by new workers, and if they succeed, that workflow will subsequently
+  only be processed by new workers. If they fail, execution goes back to old
+  workers.
+
+# Using opaque whole-worker-identifiers as worker versions
+
+Thanks to @jlegrone for his great feedback encouraging us to revive this approach.
+
+We already have a field called `binaryChecksum` (or similar) in `WorkerOptions`
+today in SDKs. This field's intent is to uniquely identify a specific worker
+build. Since we believe versioning entire workers is the easiest solution, we
+can lean on this existing field.
+
+Deprecating the name `binaryChecksum`and replacing it with something more 
+accurate like `workerBuildId` would be appropriate, since most of our SDKs
+don't even produce a binary.
+
+With this approach, users who wish to make use of the improved versioning 
+features should set the `workerBuildId` during their CI process to a value that
+is unique for that worker's implementation. _Implementation_ rather than 
+_build_ is an important distinction here, since rebuilds of identical
+code should (ideally) produce workers wit the same ID. It is not necessary to
+do this, and it is varyingly difficult depending on the language, but when
+done it enables users to produce workers with the same ID from their CI builds
+if they didn't touch worker code, but something else (ex: The CI definition itself).
+This involves some work on the part of the user, but is more reliable than expecting
+worker authors to increment a version number.
+
+Unfortunately it is infeasible to automatically set an appropriate ID for many languages.
+
+## How is versioning accomplished with the `workerBuildId`?
+
+Every new worker ID will be assumed to be incompatible with the existing id(s)
+unless explicitly specified otherwise. New workers will not process *anything* 
+until explicitly enabled if they opt into build-id based versioning. Users who 
+wish to always have new workers begin processing new workflows can make an API 
+call immediately establishing the new worker as latest as part of their 
+deployment process (see below).
+
+In order to either (safely or not) migrate *open* workflows to a new version,
+users will explicitly indicate that an ID is compatible with some already-extant
+ID.
+
+Because this behavior is a breaking change, SDKs will need to add a
+`enableBuildIdBasedVersioning` boolean flag to their `WorkerOptions`.
+
+## Server-side APIs for `workerBuildId`-based versioning
+
+Importantly, we _already_ send the id (checksum, today) on every poll request.
+It's TBD if the worker will be changed to poll a unique queue per ID, or if
+that will happen on the frontend, or matching.
+
+These APIs could very well end up in a different service, but are shown
+as part of `WorkflowService` here for simplicity.
+```protobuf
+service WorkflowService {
+    rpc SetWorkerBuildIdOrdering (SetWorkerBuildIdOrderingRequest) returns (SetWorkerBuildIdOrderingResponse) {}
```

Comment:
If I call this with `build_id=new, latest=true`, and I want `new` to no longer be latest, can I call `build_id=new, latest=false` and the previous latest will be used?

Wonder if we can automate certain types of rollback scenarios, like if a build id is failing all tasks, or not polling, auto roll back to previous latest.

Reply:
You can do it with the existing API by following the link back to the previous version, and then setting that one as latest.

But, perhaps that potentially common situation should be made easier. I don't think we want to do it exactly the way you propose since simply _not setting_ latest and explicitly setting it to false look the same to protobuf and that makes this situation confusing. 

There is already an API for marking a binary checksum as "bad" which you'd probably want to make do what you're talking about. Unfortunately that information lives only on the frontend currently and there's some internal stuff we'd have to work through, but in principle that'd be a sensible place to make do this.